docs/src/implementation.md

```diff
@@ -14,36 +14,50 @@ The parser is split into 3 modules:
 
 The IR is where semantic checking is done and where optimizations will be done in the future.
 
-A directed acyclic graph called `AlgebraicGraph` is responsible for efficiently representing all integrity constraints, identifying their type (main or auxiliary), and computing their degrees.
+A directed acyclic graph called `AlgebraicGraph` is responsible for efficiently representing all boundary and integrity constraints, identifying their domain (i.e. the rows they should be applied to), identifying their trace segment (main or auxiliary), and computing their degrees.
 
-Currently, the checks in the IR are very minimal and cover the following simple cases.
+### Building the graph
+While building the graph, the IR associates the following information with each constraint:
+- Identifies the type of the integrity constraint (transition or validity) based on the constraint expression.
+  - Constraints referencing the "next" indicator are identified as transition constraints.
+  - All other constraints are identified as validity constraints.
+- Identifies the trace segment and constraint domain (first or last row) based on the trace column and the boundary to which the constraint is applied.
```

Comment:
For clarity, maybe we should say:

```
- For integrity constraints, identifies the type (transition or validity) based on the constraint expression.
  - Constraints referencing the "next" indicator are identified as transition constraints.
  - All other constraints are identified as validity constraints.
- For boundary constraints, identifies the trace segment and constraint domain (first or last row) based on the trace column and the boundary to which the constraint is applied.
```

Reply:
Based on your suggestion, I've moved the line below these two lines:
```
- Identifies the trace segment based on the constraint expression.
  - Constraints referencing the auxiliary trace columns or using random values are identified as constraints against the auxiliary trace (trace segment 1).
  - All other constraints are identified as constraints against the main trace (trace segment 0).
```
to above and merged it with the other integrity constraints statements.

So, the overall section now reads:

```
While building the graph, the IR associates the following information with each constraint:
- For integrity constraints, identifies the trace segment and the type of the integrity constraint (transition or validity) based on the constraint expression.
  - Constraints referencing the auxiliary trace columns or using random values are identified as constraints against the auxiliary trace (trace segment 1). All other constraints are identified as constraints against the main trace (trace segment 0).
  - Constraints referencing the "next" indicator are identified as transition constraints. All other constraints are identified as validity constraints.
- For boundary constraints, identifies the trace segment and constraint domain (first or last row) based on the trace column and the boundary to which the constraint is applied.
```